examples/multicodecs/main.go

```diff
@@ -0,0 +1,181 @@
+	host "github.com/libp2p/go-libp2p-host"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ps "github.com/libp2p/go-libp2p-peerstore"
+	swarm "github.com/libp2p/go-libp2p-swarm"
+	ma "github.com/multiformats/go-multiaddr"
+
+	bhost "github.com/libp2p/go-libp2p/p2p/host/basic"
+	multicodec "github.com/multiformats/go-multicodec"
+	json "github.com/multiformats/go-multicodec/json"
+)
+
+const proto = "/example/1.0.0"
+
+// Message is a serializable/encodable object that we will send
+// on a Stream.
+type Message struct {
+	Msg    string
+	Index  int
+	HangUp bool
+}
+
+// streamWrap wraps a libp2p stream. We encode/decode whenever we
+// write/read from a stream, so we can just carry the encoders
+// and bufios with us
+type WrappedStream struct {
+	stream inet.Stream
+	enc    multicodec.Encoder
+	dec    multicodec.Decoder
+	w      *bufio.Writer
+	r      *bufio.Reader
+}
+
+// wrapStream takes a stream and complements it with r/w bufios and
+// decoder/encoder. In order to write raw data to the stream we can use
+// wrap.w.Write(). To encode something into it we can wrap.enc.Encode().
+// Finally, we should wrap.w.Flush() to actually send the data. Handling
+// incoming data works similarly with wrap.r.Read() for raw-reading and
+// wrap.dec.Decode() to decode.
+func WrapStream(s inet.Stream) *WrappedStream {
+	reader := bufio.NewReader(s)
+	writer := bufio.NewWriter(s)
+	// This is where we pick our specific multicodec. In order to change the
+	// codec, we only need to change this place.
+	// See https://godoc.org/github.com/multiformats/go-multicodec/json
+	dec := json.Multicodec(false).Decoder(reader)
+	enc := json.Multicodec(false).Encoder(writer)
+	return &WrappedStream{
+		stream: s,
+		r:      reader,
+		w:      writer,
+		enc:    enc,
+		dec:    dec,
+	}
+}
+
+// messages that will be sent between the hosts.
+var conversationMsgs = []string{
+	"Hello!",
+	"Hey!",
+	"How are you doing?",
+	"Very good! It is great that you can send data on a stream to me!",
+	"Not only that, the data is encoded in a JSON object.",
+	"Yeah, and we are using the multicodecs interface to encode and decode.",
+	"This way we could swap it easily for, say, cbor, or msgpack!",
+	"Let's leave that as an excercise for the reader...",
+	"Agreed, our last message should activate the HangUp flag",
+	"Yes, and the example code will close streams. So sad :(. Bye!",
+}
+
+func makeRandomHost(port int) host.Host {
+	// Ignoring most errors for brevity
+	// See echo example for more details and better implementation
+	priv, pub, _ := crypto.GenerateKeyPair(crypto.RSA, 2048)
+	pid, _ := peer.IDFromPublicKey(pub)
+	listen, _ := ma.NewMultiaddr(fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", port))
+	ps := ps.NewPeerstore()
+	ps.AddPrivKey(pid, priv)
+	ps.AddPubKey(pid, pub)
+	n, _ := swarm.NewNetwork(context.Background(),
+		[]ma.Multiaddr{listen}, pid, ps, nil)
+	return bhost.New(n)
+}
+
+func main() {
+	// Choose random ports between 10000-10100
```

Comment:
that's useful but then I have to use more lines to figure out which ephemeral port was picked up...